torch/csrc/jit/codegen/cuda/glfdc/eval.h

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+// Expression evaluator
+
+#include <torch/csrc/jit/codegen/cuda/glfdc/bitvector.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/expr.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/sexpr.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/sparse_map.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/stack.h>
+
+#include <algorithm>
+#include <functional>
+
+#include <c10/util/Optional.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace glfdc {
+
+struct Expr;
+struct ExprDAG;
+
+struct EvalState;
+
+// Operation for evaluator stack machine
+struct Operation {
+  const SExprRef ref;
+
+  // TODO: bitfields/ushort? 12b/16b should be plenty and we could pack it more
+  unsigned nsubops; // number of following operations to skip if lazy evaluated
+                    // (makes possible to recover tree structure)
+  unsigned noperands; // total number of operands on stack to calculate
+                      // subexpression value
+
+  static constexpr unsigned INVALID_POP = unsigned(-1);
+};
+
+using opt_index_t = c10::optional<std::size_t>;
+
+// Holds mapping of reused subexpressions
+struct ReusedExprMapping {
+  // Mapping of memoized values to slot values
+  const sparse_map slot_mapping;
+
+  // Given subexpressions returns its slot number or nullopt
+  opt_index_t slot(SExprRef ref) const noexcept {
+    return slot_mapping.find(ref.index_);
+  }
+
+  std::size_t size() const noexcept {
+    return slot_mapping.size();
+  }
+
+  bool empty() const noexcept {
+    return slot_mapping.empty();
+  }
+
+  // Creates empty mapping for eager evaluation of expression
+  static ReusedExprMapping createEagerMapping() {
+    return {sparse_map{}};
+  }
+
+  // create_lazy_mapping - Allocates mapping for reused subexpression and its
+  // value slot index
+  static ReusedExprMapping createLazyMapping(
+      const BitVector& reused_sexprs) // O(n)
+  {
+    sparse_map mapping(reused_sexprs.size());
+
+    std::size_t slot_counter = 0;
+
+    for (std::size_t i = 0; i < reused_sexprs.size(); ++i) {
+      // if subexpression is reused assign mapping
+      if (reused_sexprs[i])
+        mapping.insert(i, slot_counter++);
```

Comment:
It is generally preferred to have braces always, so:

```
if (reused_sexprs[i]) {
   mapping.insert(i, slot_counter++);
}
```

Reply:
Sure. Then, should we add .clang-format with such rule in codegen/cuda directory?

Reply:
Yeah, that would be even better.

Reply:
It seems, clang-format can't do that, clang-tidy can. Guess it'll be up to me running it pre-submit. For reference `git diff -U0 HEAD^ | clang-tidy-diff.py -fix  -checks="-*,readability-braces-around-statements" -p1`